blocklib/picoscope/Picoscope.hpp

```diff
@@ -559,399 +574,385 @@ class Picoscope : public PicoscopeBlockingHelper<TPSImpl, acquisitionMode == Acq
+                    output[i] = channel.offset + channel.scale * voltageMultiplier * static_cast<float>(driverData[i]);
+                } else if constexpr (std::is_same_v<T, gr::UncertainValue<float>>) {
+                    output[i] = gr::UncertainValue(channel.offset + channel.scale * voltageMultiplier * static_cast<float>(driverData[i]), self().uncertainty());
                 }
             }
         }
-
-        // In order to validate desired frequency before startup
-        float actual_freq;
-        convertFrequencyToTimebase(this->ps_state.handle, this->ps_settings.sample_rate, actual_freq);
-
-        return {};
     }
 
-    Error driver_disarm() noexcept {
-        if (const auto status = self().driver_stop(this->ps_state.handle); status != PICO_OK) {
-            fmt::println(std::cerr, "Stop: {}", detail::getErrorMessage(status));
-            return {status};
+    template<typename TSample>
+    std::vector<gr::Tag> processTimingTriggers(std::size_t availableSamples, std::span<TSample> samples, gr::InputSpanLike auto& timingInSpan) {
+        // Note: reference for `gr::InputSpanLike auto& timingInSpan` is needed for proper work of consumeTags method
+        std::vector<std::size_t> triggerOffsets;
+        const auto               triggerSourceIndex = self().convertToOutputIndex(trigger_source);
+        if (triggerSourceIndex != std::nullopt) {
+            triggerOffsets = findAnalogTriggers<TSample>(_channels[triggerSourceIndex.value()], samples.subspan(0, availableSamples));
         }
 
-        return {};
+        const auto nowStamp = std::chrono::high_resolution_clock::now();
+        const auto now      = std::chrono::duration_cast<std::chrono::nanoseconds>(nowStamp.time_since_epoch());
+
+        std::size_t consumeTags = 0UZ;
+        for (const auto& tag : timingInSpan.tags()) {
+            if (!tag.second.contains(gr::tag::CONTEXT.shortKey())) { // only consider timing tags
+                consumeTags++;
+                continue;
+            }
+            // TODO: Allow to match multiple events to a single trigger here and allow to timeout on events where the trigger pulse was somehow lost
+            if (triggerOffsets.size() <= _timingMessages.size()) { // all triggers have already been found
+                break;
+            }
+            _timingMessages.emplace(tag.second);
+            consumeTags++;
+        }
+        timingInSpan.consumeTags(consumeTags);
+        timingInSpan.consume(consumeTags);
+        std::vector<gr::Tag> triggerTags;
+        triggerTags.reserve(triggerOffsets.size() + 1);
+        for (const auto triggerOffset : triggerOffsets) {
+            gr::property_map timing;
+            if (_timingMessages.empty()) {
+                timing = gr::property_map{{gr::tag::TRIGGER_NAME.shortKey(), "UnknownTrigger"}, {gr::tag::TRIGGER_TIME.shortKey(), static_cast<std::uint64_t>(now.count())}, {gr::tag::TRIGGER_OFFSET.shortKey(), static_cast<std::uint64_t>(0)}};
```

Comment:
TRIGGER_OFFSET should be float
```suggestion
                timing = gr::property_map{{gr::tag::TRIGGER_NAME.shortKey(), "UnknownTrigger"}, {gr::tag::TRIGGER_TIME.shortKey(), static_cast<std::uint64_t>(now.count())}, {gr::tag::TRIGGER_OFFSET.shortKey(), 0.0f}};
```